File: 2026-02-February-GFG-POTD/February 25 - Longest Subarray with Majority Greater than K/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int longestSubarray(vector<int>& arr, int k) {
        int n = arr.size();
        unordered_map<int, int> mp;
        
        int prefixSum = 0;
        int maxLen = 0;
        
        for (int i = 0; i < n; i++) {
            // Convert to +1 / -1
            if (arr[i] > k)
                prefixSum += 1;
            else
                prefixSum -= 1;
            
            // If prefix sum > 0, whole array till i is valid
            if (prefixSum > 0) {
                maxLen = i + 1;
            } 
            else {
                // Check if prefixSum - 1 exists
                if (mp.find(prefixSum - 1) != mp.end()) {
                    maxLen = max(maxLen, i - mp[prefixSum - 1]);
                }
            }
            
            // Store first occurrence only
            if (mp.find(prefixSum) == mp.end()) {
                mp[prefixSum] = i;
            }
        }
        
        return maxLen;
    }
};

File: 2026-02-February-GFG-POTD/February 25 - Longest Subarray with Majority Greater than K/Solution.py
class Solution:
    def longestSubarray(self, arr, k):
        prefix_sum = 0
        max_len = 0
        mp = {}
        
        for i in range(len(arr)):
            # Convert to +1 / -1
            if arr[i] > k:
                prefix_sum += 1
            else:
                prefix_sum -= 1
            
            # If prefix sum > 0, whole array till i is valid
            if prefix_sum > 0:
                max_len = i + 1
            else:
                # Check if prefix_sum - 1 exists
                if (prefix_sum - 1) in mp:
                    max_len = max(max_len, i - mp[prefix_sum - 1])
            
            # Store first occurrence only
            if prefix_sum not in mp:
                mp[prefix_sum] = i
        
        return max_len
